design/volume-group-snapshot.md

```diff
@@ -0,0 +1,428 @@
+    - Users will have to label the PVCs before invoking the backup operation.
+    - This label would act as a group identifier for the PVCs to be grouped under a specific VGS.
+    - It will be used to collect the PVCs to be used for a particular instance of VGS object.
+
+- Changes to the Existing PVC ItemBlockAction plugin:
+    - Currently the PVC IBA plugin is applied to PVCs and adds the RelatedItems for the particular PVC into the ItemBlock.
+    - At first it checks whether the PVC is bound and VolumeName is non-empty.
+    - Then it adds the related PV under the list of relatedItems.
+    - Following on, the plugin adds the pods mounting the PVC as relatedItems.
+    - Now we need to extend this PVC IBA plugin to add the PVCs to be grouped for a particular VGS object, so that they are processed together under an ItemBlock by Velero.
+        - First we will check if the PVC that is being processed by the plugin has the user specified VGS label.
+        - If it is present then we will execute a List call in the namespace with the label as a matching criteria and see if this results in any PVCs (other than the current one).
+        - If there are PVCs matching the criteria then we add the PVCs to the relatedItems list.
+        - This helps in building the ItemBlock we need for VGS processing, i.e. we have the relevant pods and PVCs in the ItemBlock.
+
+- Updates to CSI PVC plugin:
+    - This is the plugin which creates VolumeSnapshots for the CSI PVCs if required.
+    - We will have to update this plugin to:
+        - Create VGS for the group of PVCs
+            - At First we will check if PVC has the user specified VGS label.
+            - If yes, then it implies it needs to be part of a VGS instance.
+            - Check if VGS exists for this PVC or not, we don't want to recreate VGS object.
+            - We can check if VGS exists by listing the VGS in the namespace using the backup UUID alongwith the label used for the VGS (we will be creating VGS with backup UUID and user specified VGS  group label) and see if the PVC is part of any of the existing VGS for the particular backup operation we are currently in.
+            - If not then create the VGS object with the required labels (in this case backup UUID and the user specified VGS group label) and also add the PVC names (that are included in the VGS) as annotations. (This will be useful for our VGS restore operation where we would need PVC resource identifiers).
+            - Once VGS is created, add the VGS as an additional item to be included in the backup.
+        - Skip creation of VS for PVC that are already included in VGS
+            - Now if the PVC has the user specified VGS label that means it is destined to be part of some VGS and the VS would be created by the K8s VGS controller and Velero does not need to explicitly create the VS for the PVC.
+
+- Add a new VolumeGroupSnapshot(VGS) BIA plugin
```

Comment:
Another thing for data mover:
If we do nothing, the data mover modules don't know the existence of VGS objects, but they know VS/VSC.
The VS/VSC is deleted by the data mover modules after the data movement completes, then who is going to delete the VGS objects?

This leads to a question: can we delete the VGS objects after the individual VS/VSC are created? If yes, then we will have a simple solution.

Reply:
"This leads to a question: can we delete the VGS objects after the individual VS/VSC are created? If yes, then we will have a simple solution."
That's a good question. This will probably need to be determined while testing during impl (or prior to impl). But yes, we need to figure out the right mechanism for making sure that VGS *and* VGSC get deleted at the appropriate time.

Reply:
Yes, we probably need to do a POC particularly for this case and decide how to make the solution.

Reply:
Data mover needs to visit the VS object, so if VGS and VS are still binding together (deleting one causing deleting the other) even after the snapshot is created, that is a bad news.

Reply:
As the current behavior, data mover will clean up VS and VSC.
For the VGS case, if the VGS and VGSC still exists, can we still delete the VS/VSC by data mover before finalizing phase?

Reply:
@Lyndon-Li I don't think we can delete the VS which is part of the VGS directly. The VS has protection via finalizer: 

```yaml
  finalizers:
  - snapshot.storage.kubernetes.io/volumesnapshot-in-group-protection
```

Reply:
The data mover snapshot exposer does some significant transformations (retain->delete) to the VS/VSC delivered to it. So I am afraid we have to make them independent from VGS/VGSC.

Reply:
> I don't think we can delete the VS which is part of the VGS directly

No, the question is whether we can delete the VGS/VGSC while VS/VSC still exist

Reply:
Thats only possible if the `deletionPolicy` is `Retain` for VGSClass.

Reply:
What if we change VGSC's deletionPolicy before deleting it? Just as same as what we do for VSC.

Reply:
I think we can probably make a common workflow for both data mover and non data mover case:
1. create VGS
2. Wait VGSC
3. Wait VS/VSC for each volume
4. Wait snapshot handle for each volume
5. Set VGSC to `Retain`
6. Delete VGS and VGSC
7. Let go into the existing async operations (different for data mover and non data mover)

@shubham-pampattiwar @sseago @blackpiglet

Reply:
@Lyndon-Li That looks right to me.